### PR TITLE
Improve migration/replication system 

### PR DESCRIPTION
**Pull request by mdogan**
This work improves migration/replication system to makes it safe against data loss. 

For design details please see: https://hazelcast.atlassian.net/wiki/display/COM/Avoid+Data+Loss+on+Migration+-+Solution+Design

_Additional details can be seen in individual commit messages._

Fixes https://github.com/hazelcast/hazelcast/issues/5444

EE counter-part: https://github.com/hazelcast/hazelcast-enterprise/pull/850

Q. E. D.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/instance/Node.java [hunk 1]
```diff
@@ -259,6 +260,13 @@ private void initializeListeners(Config config) {
                 known = true;
             }
 
+            if (listener instanceof InternalMigrationListener) {
+                final InternalPartitionServiceImpl partitionService =
+                        (InternalPartitionServiceImpl) nodeEngine.getPartitionService();
+                partitionService.setInternalMigrationListener((InternalMigrationListener) listener);
```
Comment (gurbuzali) [inline hazelcast/src/main/java/com/hazelcast/instance/Node.java]: what if there are more than one internal listener, latter one will override ?

Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/instance/Node.java]: This is not user api or spi. It is only done for testing. We use InternalMigrationListener to execute a specific test scenario deterministically within our integration tests so we register only a single internal listener and perform the test with it.

**Comment by metanet** [comment]
This PR has received yet another conflict because of yet another refactoring in the middle of the release...

**Comment by metanet** [comment]
I will not wait for another 8 days for the review. After the conflict is solved, I will merge the PR if there is no review for a couple of following days.

**Comment by gurbuzali** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java [hunk 2]
```diff
@@ -54,8 +61,33 @@ public int getPartitionId() {
...
+        return currentReplicaIndex;
+    }
+
+    /**
+     * Gets the index of the partition replica that current member will own after migration is committed.
+     * This index will be -1 if partition replica will be moved from current member completely.
+     *
+     * @return index of the partition replica that current member will own after migration is committed
+     */
+    public int getNewReplicaIndex() {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java]: could you please describe how the `MigrateAwareService` is supposed to use this information?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/ExecutionServiceImpl.java [hunk 3]
```diff
@@ -167,8 +167,7 @@ private ManagedExecutorService createExecutor(String name, int poolSize, int que
             Node node = nodeEngine.getNode();
             String internalName = name.startsWith("hz:") ? name.substring(BEGIN_INDEX) : name;
             HazelcastThreadGroup hazelcastThreadGroup = node.getHazelcastThreadGroup();
-            PoolExecutorThreadFactory threadFactory = new PoolExecutorThreadFactory(hazelcastThreadGroup,
-                    hazelcastThreadGroup.getThreadPoolNamePrefix(internalName));
+            PoolExecutorThreadFactory threadFactory = new PoolExecutorThreadFactory(hazelcastThreadGroup, internalName);
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/ExecutionServiceImpl.java]: well spotted!

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    private Future invoke(Operation operation, int replicaIndex, int partitionId) {
+        OperationService operationService = nodeEngine.getOperationService();
+        return operationService.createInvocationBuilder(SERVICE_NAME, operation, partitionId)
+                .setTryCount(INVOCATION_TRY_COUNT)
+                .setTryPauseMillis(INVOCATION_TRY_PAUSE_MILLIS)
+                .setReplicaIndex(replicaIndex)
+                .invoke();
+    }
+
+    // TODO: VISIBILITY PROBLEM! Replica versions are updated & read only by partition threads!
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: can you please elaborate the problem?

Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: This will be fixed by graceful shutdown work. Currently, this part reads replica versions from a non-partition thread without guaranteeing thread safety. Normally they are accessed only within partition threads.

File: hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java (hunk 2 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java]: `MigrateAwareService` should decide to keep or remove, all or some part of its data, by looking `currentReplicaIndex`, `newReplicaIndex` and `migrationEndpoint`.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java (hunk 4 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: just to double-check I understand it properly - right now a given `partitionId` is always access by a same thread -> there is no visibility problem. Is this correct?

The warning serves for other developers to avoid touching this from other threads, am I right?

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: This problem is NOT introduced by this PR, it's already there in the current master. We just spotted it and added a TODO. Also it's already fixed by our graceful shutdown work, as Basri mentioned.

File: hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java (hunk 2 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java]: do you think some sort of guidance help to implementers? 
by looking at our current implementations the logic seem to be quite similar.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java (hunk 4 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: that's fine, I'm trying to understand it, that's all.

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: @jerrinot: correct. normally replica-version array is accessed/modified by only it's specific partition thread. But usage here, in the method below the TODO, is not thread-safe. Version array is accessed out side of partition threads.

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java]: ✓ thanks!

File: hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java (hunk 2 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java]: Logic is similar. But now there are more parameters than before. I can add an example to the class  (`MigrateAwareService`) javadoc.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalMigrationListener.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+package com.hazelcast.internal.partition.impl;
+
+
+import com.hazelcast.internal.partition.MigrationInfo;
+
+import java.util.EventListener;
+
+public abstract class InternalMigrationListener implements EventListener {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalMigrationListener.java]: JavaDoc explaining the purpose (=deterministic testing) of this class would be nice.

File: hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java (hunk 2 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/PartitionMigrationEvent.java]: an example would be excellent! 

before the logic was quite simple: a sending side clears its partition on commit, receiving side clears it on rollback. that's very intuitive. 

now you need to take backup count into consideration.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    private static final boolean TRACE = false;
+
+    interface MigrationDecisionCallback {
+        void migrate(Address source, int sourceCurrentReplicaIndex, int sourceNewReplicaIndex, Address destination,
+                     int destinationCurrentReplicaIndex, int destinationNewReplicaIndex);
+    }
+
+    private final Set<Address> verificationSet = new HashSet<Address>();
+
+    // checkstyle warnings are suppressed intentionally because the algorithm is followed easier within a single coherent method.
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java]: I understand your motivation behind this decision, but could you include comment on individual cases/branches? 
It's quite hard to follow it unless you have a good understanding of possible migrations cases.

Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java]: We have test cases for that

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java]: I am aware of that, but having it in-lined in the code would really readability. 

This is usually achieved by having descriptive method names. However you decided for a single method handling all cases - I understand your choice, but one can feel kinda lost. 

A few words would do it. there is already a comment like this. see `// IT IS A SHIFT DOWN`

Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java]: I am still not into it but we can add more comments anyway.

Imho, the whole algorithm is not very easy to follow by only checking the code. If someone really wants to grasp the algorithm completely, he should check the test cases and the design document. We have already provided the full description of the algorithm there.

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java]: I'll add more comments and javadoc. just waiting for review to complete.

**Comment by jerrinot** [comment]
:+1: some minor comments.
It's truly an epic work, hats off to gentlemen!

**Comment by bwzhang2011** [comment]
@mdogan, would you mind sparing some time and take a look at issue https://github.com/hazelcast/hazelcast/issues/7987, could the latest improvement for migration solve such problem ?
